src/lem/circuit.rs

```diff
@@ -0,0 +1,889 @@
+pub struct SlotsInfo {
+    slots_map: IndexMap<LEMOP, (usize, SlotType)>,
+    slots: IndexSet<(usize, SlotType)>,
+}
+
+impl SlotsInfo {
+    #[inline]
+    pub fn get_slot(&self, op: &LEMOP) -> Result<&(usize, SlotType)> {
+        self.slots_map
+            .get(op)
+            .ok_or_else(|| anyhow!("Slot not found for LEMOP {:?}", op))
+    }
+}
+
+impl LEMOP {
+    /// STEP 1: compute the slot mapping on a first (and unique) traversal
+    ///
+    /// While traversing parallel paths, we need to reuse compatible slots. For
+    /// example, on a `MatchTag` with two arms, such that each arm is a `Hash2`,
+    /// we want those hashes to occupy the same slot. We do this by using a
+    /// `SlotsCounter`, which can return the next index for a slot category
+    /// (`Hash2`, `Hash3` etc).
+    ///
+    /// Further, when we:
+    /// * Construct the same pointer twice
+    /// * Destruct the same pointer twice
+    /// * Construct a pointer that was previously destructed
+    /// * Destruct a pointer that was previously constructed
+    ///
+    /// We want to reuse the same slot as before. To accomplish this, we use
+    /// hashmaps that can recover the slots that were previously allocated.
+    pub fn slots_info(&self) -> Result<SlotsInfo> {
+        let mut slots_map = IndexMap::default();
+        let mut slots = IndexSet::default();
+
+        // these hashmaps keep track of slots that were allocated for preimages
+        let mut preimgs2_map: HashMap<&[MetaPtr; 2], (usize, SlotType)> = HashMap::default();
+        let mut preimgs3_map: HashMap<&[MetaPtr; 3], (usize, SlotType)> = HashMap::default();
+        let mut preimgs4_map: HashMap<&[MetaPtr; 4], (usize, SlotType)> = HashMap::default();
+
+        // these hashmaps keep track of slots that were allocated for images
+        let mut imgs2_map: HashMap<&MetaPtr, (usize, SlotType)> = HashMap::default();
+        let mut imgs3_map: HashMap<&MetaPtr, (usize, SlotType)> = HashMap::default();
+        let mut imgs4_map: HashMap<&MetaPtr, (usize, SlotType)> = HashMap::default();
+
+        let mut slots_counter = SlotsCounter::default();
+        let mut stack = vec![(self, Path::default())];
+        while let Some((op, path)) = stack.pop() {
+            /// Designates a slot for a pair of preimage/image. If a slot has
+            /// already been allocated for either the preimage or the image,
+            /// reuses it. Otherwise, allocates a new one.
+            macro_rules! populate_slots_map {
+                ( $preimg: expr, $img: expr, $preimgs_map: expr, $imgs_map: expr, $counter_fn: expr ) => {
+                    match ($preimgs_map.get($preimg), $imgs_map.get($img)) {
+                        (Some(slot), _) | (_, Some(slot)) => {
+                            // reusing a slot
+                            if slots_map.insert(op.clone(), *slot).is_some() {
+                                bail!("Duplicated LEMOP: {:?}", op)
```

Comment:
This *particular* check won't ensure that the same hash with different *tags* won't be duplicated. Please confirm that is ensured elsewhere and this is just a supplemental check.

Reply:
Related thought (remember, I am just trying to first understand the algorithm):

If both img and preimg match, then we need to be sure that they match the *same* slot, right? That case is not checked. Is it guaranteed that this will be the case?

Reply:
If preimage and image match, it's a malformed LEM
```rust
let x: Cons = hash2(a, b)
let x: Sym = hash2(a, b)
```
is not in SSA

```rust
match_tag k
  foo => let x = hash2(a, b)
  bar => let x = hash2(a, b)
```
Didn't go through `deconflict`

Reply:
Taking the first example, please walk me through it. Is this malformed and therefore will have been prevented?
```
let x: Cons = hash2(a, b)
let y: Sym = hash2(a, b)
```
If not, will the img and preimg not be the same?

If they are the same, what does that imply? I think the answer is that this is fine — they can match, and they should consequently use the same slot. However, it *must* be the case that their tags differ (because otherwise… malformed).

But, that seems to contradict what you wrote above (that they cannot both match). So please walk me through that case and why they cannot both match and share a slot.

Reply:
It's not malformed.
After interpretation, `x` and `y` will be bound to different `Ptr`s because they have different tags.
They will use the same slot because the preimage is the same and they're both `LEMOP::Hash2`.
However, `x` will be constrained to have a `Cons` tag and `y` will be constrained to have a `Sym` tag

Reply:
Okay, so they will have different tags, as expected. But I don't see how the tags are being taken into account at all. For example, in the `match op { … }` below — where `populate_slots_map!` is being used, the tags are completely ignored.

Maybe I'm just misunderstanding how this is supposed to work. What about the case in which you perform both a hash and unhash on the same img/preimg pair? I'm guessing at the syntax:
```
let x: Cons = hash2(a, b);
let (a, b) = unhash2(x);
```
Is that malformed? If not, don't we expect both img and preimg to match?

Reply:
Answering my own question: I think this example *is* malformed because of SSA.

Reply:
Okay, I think I get the previous example. `x` and `y` don't match because their types differ, and these are pointer comparisons. Is that it?

Reply:
For clarity, `hash(a, b)` will occupy the same slot both when declaring `x` and when declaring `y`. Wen you declare a new meta pointer with `let x: TAG = hashX(...)`, the `TAG` is irrelevant to the slot allocation. In this case, the slot is about making sure that the hash of `x` equals the hashing of the pointers fed to `hashX`

Reply:
I'm not sure what you mean by "`x` and `y` don't match"

Reply:
Okay, still thinking out loud, what about different branches?

First,
```
let y = …
```
Then in one branch (using a hypothetical future equality LEMOP):
```
let x: Cons = hash(a, b);
let z = (x == y);
```
and in another branch:
```
let (a, b) = unhash2(y);
```
I think this would not be malformed. Does that mean we will definitely never have such a LEMOP? If so, will that conflict with implementing equality checks?

Reply:
Mmmmm, this example is still nonsense.

Reply:
In the example above, we will have one `Hash2` slot because we hit a memoized preimage `[a, b]`

Reply:
The discussion about deduplication is now out of the scope of this PR